src/BloomExe/web/EnhancedImageServer.cs

```diff
@@ -1,51 +1,137 @@
 using System;
 using System.Diagnostics;
-using System.Drawing.Text;
 using System.Globalization;
-using System.Linq;
-using System.Net;
-using Bloom.ImageProcessing;
 using System.IO;
-using L10NSharp;
+using System.Threading;
 using Microsoft.Win32;
+using L10NSharp;
 using Palaso.IO;
 using Bloom.Collection;
-
+using Bloom.ImageProcessing;
 
 namespace Bloom.web
 {
 	/// <summary>
 	/// A local http server that can serve (low-res) images plus other files.
 	/// </summary>
+	/// <remarks>geckofx makes concurrent requests of URLs which this class handles. This means
+	/// that the methods of this class get called on different threads, so it has to be
+	/// thread-safe.</remarks>
 	public class EnhancedImageServer: ImageServer
 	{
 		private FileSystemWatcher _sampleTextsWatcher;
 		private bool _sampleTextsChanged = true;
+		private const int RWLockTimeOut = 100;
 
 		public CollectionSettings CurrentCollectionSettings { get; set; }
 
 		public EnhancedImageServer(LowResImageCache cache): base(cache)
 		{
+			Lock = new ReaderWriterLock();
+		}
+
+		// We use three different locks to synchronize access to the methods of this class.
+		// This allows certain methods to run concurrently.
+
+		// used to synchronize access to properties. Use this lock when you want to allow
+		// multiple concurrent readers, but no one else when writing the property.
+		private ReaderWriterLock Lock { get; set; }
+		// used to synchronize access to I18N methods
+		private object I18NLock = new object();
+		// used to synchronize access to various methods
+		private object SyncObj = new object();
+
+		private string _currentPageContent;
+		private string _accordionContent;
+		private bool _authorMode;
+
+		public string CurrentPageContent
+		{
+			get { return ReaderLock(() => _currentPageContent); }
+			set { WriterLock(() => _currentPageContent = value); }
 		}
 
-		public string CurrentPageContent { get; set; }
-		public string AccordionContent { get; set; }
-		public bool AuthorMode { get; set; }
+		public string AccordionContent
+		{
+			get { return ReaderLock(() => _accordionContent); }
+			set { WriterLock(() => _accordionContent = value); }
+		}
+
+		public bool AuthorMode
+		{
+			get { return ReaderLock(() => _authorMode); }
+			set { WriterLock(() => _authorMode = value); }
+		}
 
 		/// <summary>
-		/// There can really only be one of these globally, since ReadersHandler is static. But for now that's true anyway
-		/// because we use a fixed port. See comments on the ReadersHandler property.
+		/// There can really only be one of these globally, since ReadersHandler is static. But for
+		/// now that's true anyway because we use a fixed port. See comments on the ReadersHandler
+		/// property.
 		/// </summary>
 		public Book.Book CurrentBook
 		{
-			get { return ReadersHandler.CurrentBook; }
-			set { ReadersHandler.CurrentBook = value; }
+			get
+			{
+				lock (SyncObj)
```

Comment:
It's no obvious why this isn't a Reader/Writer lock situation...because the ReadersHandler might change? A comment might be helpful.